modules/core/src/main/scala/ch/epfl/bluebrain/nexus/kg/core/resources/Resources.scala

```diff
@@ -84,22 +88,41 @@ class Resources[F[_], Type <: ResourceType](agg: Agg[F], project: Project)(impli
   /**
     * Attempts to add an attachment to a resource.
     *
-    * @param id     the identifier of the resource
-    * @param schema the identifier of the schema that this resource validates against
-    * @param rev    the last known revision of the resource instance
-    * @param value  the attachment metadata
-    * @param tags   the tags associated to this operation
+    * @param id            the identifier of the resource
+    * @param schema        the identifier of the schema that this resource validates against
+    * @param rev           the last known revision of the resource instance
+    * @param wrappedSource the attachment source + its metdata
+    * @param tags          the tags associated to this operation
     * @return a [[IdVersioned]] instance wrapped in the abstract ''F[_]'' type
     *         if successful, or a [[ch.epfl.bluebrain.nexus.kg.core.rejections.Fault]] wrapped within ''F[_]'' otherwise
     */
-  def attach(id: String, schema: String, rev: Long, value: Attachment, tags: Set[String] = Set.empty)(
+  def attach(id: String, schema: String, rev: Long, wrappedSource: SourceWrapper[In], tags: Set[String] = Set.empty)(
       implicit caller: CallerCtx,
-      @silent access: Type HasAccess Attach): F[IdVersioned] =
-    for {
-      _ <- projectUnlocked()
-      state <- eval(Cmd.Attach(reprId(id, schema), rev, caller.meta, value, tags + project.id),
-                    s"Attach resource '$id'")
-    } yield IdVersioned(state.id, state.rev)
+      @silent access: Type HasAccess Attach): F[IdVersioned] = {
+    val repr = reprId(id, schema)
+    val inOutResult: F[Either[Rejection, Attachment]] = F.pure {
+      logger.debug(
+        s"Uploading file '${wrappedSource.filename}' with mediaType '${wrappedSource.mediaType}' to instance '$id'")
+    } flatMap { _ =>
+      projectUnlocked()
+    } flatMap { _ =>
+      agg.currentState(repr.persId).map(State.eval(_, Cmd.Attach(repr, rev, caller.meta, EmptyAttachment)))
```

Comment:
why do we need this evaluation before writing the binary to disk? and why do we use EmptyAttachment?